token/program-2022/src/extension/reallocate.rs

```diff
@@ -83,8 +84,23 @@ pub fn process_reallocate(
     )?;
 
     // unpack to set account_type, if needed
-    let mut token_account = token_account_info.data.borrow_mut();
-    set_account_type::<Account>(&mut token_account)?;
+    let mut token_account_data = token_account_info.data.borrow_mut();
+    set_account_type::<Account>(&mut token_account_data)?;
+
+    // sync the rent exempt reserve for native accounts
+    let mut token_account = StateWithExtensionsMut::<Account>::unpack(&mut token_account_data)?;
+    if token_account.base.is_native.is_some() {
+        let new_amount = token_account_info
+            .lamports()
+            .checked_sub(new_minimum_balance)
+            .ok_or(TokenError::Overflow)?;
+        if new_amount < token_account.base.amount {
+            return Err(TokenError::InvalidState.into());
+        }
```

Comment:
hmm, i think this cause unwanted error for this edge case:

1. prepare native token account without extensions (account data size: 165) with `token_account_info.lamports() == 2_081_040` (= rent exempt minimum for account data size: 171). so `token_account.base.amount == 41_760` (`2_081_040` - `2_039_280`).
2. call `reallocate(..., &[ExtensionType::CpiGuard], ...)`, this reallocates to be +6 more bytes: account data size: 171
3. so this `if` evaluates to `if 0 < 41_760` and this triggers error. i think this instead succeed.

Reply:
Yeah that's true, I was thinking they should do a "sync native" beforehand, but there's really no reason to

Reply:
So one thing that seems a little weird to me now is that as-written, the `reallocate_syncs_native::transfer_lamports_only` magically syncs the additional lamports so that post `base.amount = 999958240`. But if you call the same transfer+reallocate with no extensions (empty slice), there is no sync and the post `base.amount` is still 0.

Maybe this is the behavior we want, but how about a test case that makes this clear?
...I'm realizing we should have had tests that cover realloc with empty slice in the first place.

It would also be nice to have test cases that cover transfers smaller than the diff amount needed for the extension.

Reply:
That's a good point -- we can skip the sync native bit, and only update the rent exempt reserve. What do you think?

Reply:
I got convinced that was better, and added some empty tests too.

Reply:
yeah, i think so. actually syncing is kind of _transfer_ and we would need to account for existence of CpiGuard extension in very broad sense. by not syncing, we're avoiding altogether this semantic war. :)